include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,382 @@
+			{
+				inImg,
+				outImg,
+				inParams,
+				outParams,
+				inParams.format,
+				outParams.format,
+				getTexelOrBlockBytesize(inParams.format),
+				getTexelOrBlockBytesize(outParams.format),
+				reinterpret_cast<const uint8_t*>(inImg->getBuffer()->getPointer()),
+				reinterpret_cast<uint8_t*>(outImg->getBuffer()->getPointer()),
+				inImg->getRegions(state->inMipLevel),
+				outRegions,
+				outRegions.begin(), {}, {}
+			};
+
+			auto allocateCache = [](core::smart_refctd_ptr<asset::ICPUBuffer> cache, size_t size)
+			{
+				auto newPointer = reinterpret_cast<uint8_t*>(_IRR_ALIGNED_MALLOC(size, _IRR_SIMD_ALIGNMENT));
+				cache = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(size, newPointer, core::adopt_memory);
+			};
+
+			allocateCache(state->rowSumCache, outExtent.X * texelByteSize);
+			allocateCache(state->columnSumCache, outExtent.Y * texelByteSize);
+
+			auto perOutputRegion = [state]( const CMatchedSizeInOutImageFilterCommon::CommonExecuteData& commonExecuteData, CBasicImageFilterCommon::clip_region_functor_t& clip) -> bool
+			{
+				const auto blockDims = asset::getBlockDimensions(inFormat);
+
+				auto sum = [&commonExecuteData, &blockDims](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
+				{
+					auto localOutPos = readBlockPos * blockDims + commonExecuteData.offsetDifference; 
+					auto outPtrTexelPosition = commonExecuteData.outData + commonExecuteData.oit->getByteOffset(localOutPos, commonExecuteData.outByteStrides);
+					auto inPtrTexelPosition = commonExecuteData.inData + readBlockArrayOffset;
+
+					constexpr auto MAX_PLANES = 4;
+					const void* srcInPix[MAX_PLANES] = { inPtrTexelPosition, nullptr, nullptr, nullptr };
+					uint8_t* dstPix = commonExecuteData.outData + commonExecuteData.oit->getByteOffset(localOutPos, commonExecuteData.outByteStrides);
+
+					// with caching we can just substract one value (looking at left or down) to get appropriate sum without summing everything many times per texel 
+
+					// fill cache with max values
+					auto fillCacheIfEnteredMaxValues = [&]()
+					{
+						const auto width = commonExecuteData.oit->imageExtent.width;
+						const auto height = commonExecuteData.oit->imageExtent.height;
+						core::vector4df_SIMD channels; // tmp, TODO
+
+						//convertColor<inFormat, outFormat>(srcPix, dstPix, blockX, blockY); nah		
+
+						if (localOutPos.X == width - 1 && localOutPos.Y == height - 1)
+						{
+							double finalMaxSumForTexel = {};
+
+							// encode and decode it properly
+							for (size_t yPos = 0; yPos < height; ++yPos)	// compute max sum values in rows for y texel sum values
+							{
+								// use finalMaxSumForTexel and pointer arythmetic to calculate final sum for a texel, towards LEFT
+								// gain max sum for X values per each y
+							}
+
+							for (size_t xPos = 0; xPos < width; ++xPos)
```

Comment:
cache coherence, first sweep x, then sweep y

Reply:
it doesn't matter anymore